### ISSUE TITLE
Creating a Record, setting no fields explicitly, doing .insert(), PK is not set into Record object

### ISSUE THREAD

**Issue by adrianmsmith** [jOOQ/jOOQ#10282]
Related to #6546

### Expected behavior and actual behavior:

**Expected behavior:** A database table has an auto-generated PK and other nullable fields. Create a Record object and do a `record.insert()`. The database assigns a PK value to the newly-inserted row and the .insert() sets that value into the Record object.

**Actual behaviour:** If you set 1 or more fields explicitly (even setting them to null), the PK value is written correctly into the Record object during the .insert(). If you set 0 fields explicitly, the PK is _not_ written into the Record object during the .insert(). The .insert() does not throw an exception, get `record.getId()` is null afterwards.

### Steps to reproduce the problem (if possible, create an MCVE: https://github.com/jOOQ/jOOQ-mcve):

https://github.com/adrianmsmith/jooq-insert-record-no-fields-no-pk

### Versions:

- jOOQ: 3.13.2
- Java: 11
- Database (include vendor): H2 (or PostgreSQL)
- OS: macOS 10.14.6
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thank you for your report. The reason why it is not set is because the record is not inserted, see also: https://github.com/jOOQ/jOOQ/issues/6935

This works by (flawed) design. It will be difficult to change as users may rely on the current no-op behaviour.

Closing this as a duplicate of #6935

**Comment by lukaseder** [comment]
Looking into this now. There's a small difference between `insert()` and `store()`/`update()`, which leads to an irregularity. While there is a statement to insert a default record, namely:

```sql
INSERT INTO t DEFAULT VALUES;
```

(or an emulation of the above) ... no such thing exists for `UPDATE` statements. If an `insert()` call would guarantee insertion of the record, with all the side effects associated with it (generation of default values, triggers, etc.), then it could be reasonably expected for `update()` calls to do the same, e.g. to trigger updates of `LAST_MODIFIED` columns by triggers. A hack would be to just update any random column to use itself as a value:

```sql
UPDATE t SET x = x WHERE id = :id;
```

But which one to pick? The first? The primary key? All columns? We could argue that this feature should only be implemented for `insert()`, not for `update()`, but then `store()` would behave differently depending on whether the record is new (`insert()`) or fetched (`update()`).

What about the new `merge()` method? It produces an `INSERT .. ON DUPLICATE KEY UPDATE` statement without a `SET` clause. Probably a bug, but what else could we do?

What about records on tables where there is no identity, typically when they have a natural key? The primary key cannot be generated, so it must be supplied. These `insert()` calls currently don't execute. Changing this would be a breaking change.

I think the workaround here to just run the `INSERT INTO t DEFAULT VALUES` statement explicitly, or set individual columns, is the best thing we can do without adding new problems and irregularities, or even regressions.

This isn't strictly a duplicate of https://github.com/jOOQ/jOOQ/issues/6935, which is about DAOs, where there are no notions of `changed` flags as in records.

**Comment by lukaseder** [comment]
We could consider a new `Settings` flag that governs this. But then, does that flag just skip the "executable" test and append `DEFAULT VALUES` instead? (1) Or will it have the semantics of setting every `changed` flag to `true`? (2)

(1) would just address this special case of wanting to generate a surrogate key with no data. In a strict relational sense, this has no meaning. The only reason why anyone would want to do this is merely technical (e.g. to "reserve" the space for later updates). The insertion of such an emtpy record can probably be delayed, too, without any penalties to business logic (although there might be purely technical hurdles).

(2) would address a more general problem, where the underlying data might have changed, and we do not want to implement optimistic locking, but just write over the changes, regardless if our version has changed as well. This would then work for both `INSERT`, `UPDATE`, or `MERGE` statements.

Of course, both of these things can be solved by writing down the query explicitly, so they're merely convenience, not necessary. I'm undecided, here. I will keep this open in case someone else requires this as well, and can provide use-cases that justify such a change.

**Comment by lukaseder** [comment]
Getting more feedback on this topic from twitter:
https://twitter.com/lukaseder/status/1281600147575775232

**Comment by lukaseder** [comment]
Interesting input that I haven't even thought of yet:
https://twitter.com/radityopw/status/1281601315580162049

Executing the `INSERT` as it is makes sense only if *all* non-primary key columns are nullable. This is of course the case in your case, but quite unlikely in an average schema.

**Comment by adrianmsmith** [comment]
I think the latter point about nullable fields isn’t quite correct. Other
columns could have DEFAULTs. Maybe it’s splitting hairs.

I understand the point about triggers. I hadn’t thought of that.
Nevertheless, even if the behavior between INSERT and UPDATE were
different, and that’s not ideal, the current situation isn’t ideal either.
So the question is, which is less ideal.

Not everyone who is using databases is using triggers. But everyone who
does a `r.insert()` expects a row to be inserted (maybe I’m wrong about
that, and it’s just me, let’s see what the twitter poll says).

What I mean is, I would say insert/update inconsistent is better than
`r.insert()` not inserting anything. But I understand this might be just my
opinion.

My use-case was: I have a user table and various other tables storing
various things regarding the user such as permissions. The user table just
has (user_id, username, password_hashed). When you click “free trial” of
the SaaS you should go right into using the software. No need to enter u/p
just yet. So those cols were null. But rows in the other tables
(permissions) need to be written, the user needs to exist, thus the insert.

I agree that the workaround is trivial. The main problem this caused me
wasn’t that I couldn’t get it to work once I realized what was going on, it
was that it took me an hour of debugging to work out what was going on :)
Once I realized the behavior the workaround was easy. So my issue was more
it didn’t “just work”. Although maybe I didn’t study the docs closely
enough.

`merge()` is a difficult one ...

I’m not too bothered about breaking changes, but that’s just me. I expect
breaking changes in major versions e.g. the handling of Instant and
TIMESTAMPTZ in a recent version.

**Comment by lukaseder** [comment]
> I think the latter point about nullable fields isn’t quite correct. Other columns could have DEFAULTs. Maybe it’s splitting hairs.

Not splitting hairs, yes the defaults would apply. So, the assumption is that all columns are any of:

- The identity *and* primary key
- Nullable
- Defaulted

> Not everyone who is using databases is using triggers. But everyone who does a `r.insert()` expects a row to be inserted (maybe I’m wrong about that, and it’s just me, let’s see what the twitter poll says).

So far, it's an unrepresentative (34 votes) majority of 87% (65% in favour of changing `insert()` behaviour only, + 22% in favour of changing both). But there are also those who expect the status quo (8% they'll run into the regression).

> What I mean is, I would say insert/update inconsistent is better than `r.insert()` not inserting anything. But I understand this might be just my opinion.

An inconsistency needs much more documentation, needs much more diligence both on the maintainer and user side, causes much more cognitive friction, and introduces new risks of oversights of even weirder edge cases... You probably don't really care about the `update()` behaviour, you just want the `insert()` behaviour fixed, and that's understandable. But the inconsistency is a high price to pay for the rush, so, I'm still hoping that a more thorough solution pops up.

In any case, since this would introduce regressions, there will be `Settings`, and since people don't necessarily agree on whether consistency is needed, nothing prevents us from adding more than one setting,

- One for the `INSERT` behaviour
- One for the `UPDATE` behaviour (possibly even including the strategy of how to emulate the `UPDATE` behaviour, i.e. touch PK, touch all columns, etc.)

> My use-case was [ ... ]

Thanks for sharing that. Indeed, a technical workaround to what is a tricky modelling problem.

Clearly, a username shouldn't be null because it's a user candidate key, along with their email address, but obviously, a user is a resource that can exist prior to being complete. Within a transaction, there are deferrable constraints (in some RDBMS), which help here, but you probably want this interaction to go over several transactions. I think I would have created a UUID or some other generated value with low collision probability for username / password. Technically, this is no different from allowing the user to change their user name or password later on.

Anyway. It's a good example of what keeps happening with sequence generated surrogate keys as opposed to UUIDs or natural keys. There are probably many other users who have tables with "reserved" records like that.

> So my issue was more it didn’t “just work”. Although maybe I didn’t study the docs closely enough.

It's documented (which makes the change even less backwards compatible), but "just work" is a reasonable expectation. Outside of these silly `UpdatableRecords`, jOOQ tends to "just work". It's the `UpdatableRecords` that keep causing annoying edge cases, and this is definitely one.

Another one is when loading a POJO into a record, all values are flagged as `changed` (because they're "set"). This may not be desireable when setting a `NULL` value from a POJO into a record, especially not when it's the primary key (which is `NOT NULL` and has an identity / `DEFAULT` value). So, we're not setting the primary key's `changed` flag in that case. It's the right thing to do to "just work", but it's really complicated.

The difference to this case here is that in the other case, there was an exception before, which was never "right", and the improvement really is an improvement. It's hard to imagine a case where someone relied on an exception caused by a `NULL` value in a primary key column. Here, people may rely on the status quo, including the resulting update count of zero.

> I’m not too bothered about breaking changes, but that’s just me. I expect breaking changes in major versions e.g. the handling of Instant and TIMESTAMPTZ in a recent version.

Think about the hour you wasted. This translates to several hours wasted on a regression, because in your case, you were trying to do something that didn't work, but you were actively thinking about it. Someone upgrading from jOOQ 3.8 to 3.14 would spend many hours going through release notes to figure out what the breaking change could be.

But perhaps, the amount of person hours spent on this particular future regression are less than the amount of person hours spent on the status quo...

**Comment by adrianmsmith** [comment]
> possibly even including the strategy of how to emulate the UPDATE behaviour, i.e. touch PK, touch all columns, etc.) 

Yes the problem of there not being an "UPDATE this row but don't change anything" in SQL makes it really difficult to come up with an elegant solution for executing a database statement for `r.update()` in the case nothing has changed :-(

I suppose when SQL was designed they didn't think of triggers. An INSERT with everything default changes the data, but an UPDATE with nothing changed doesn't. So without triggers there is no need for a "no-change" UPDATE. But triggers are a thing...

> Clearly, a username shouldn't be null because it's a user candidate key

Perhaps a separate table with the username/email/password attributes, which could either have a row per user (for those with completed registration) or not have a row (for those users without completed registration), might have been a better solution than nullable columns, in retrospect. But in that case, the user table would just have one column (user_id) and the same problem would apply.

I'm not convinced about using fake data for the username. Inevitably at one point or another this is going to leak through to the user due to some bug or oversight and they'll not understand why their username is some garbage string.

And to determine "has this user fully registered yet" (e.g. to show a prompt banner on the screen encouraging them to do so) is going to be more difficult with fake data. Perhaps requiring an extra field. I think null or having a separate table for the data is a more elegant solution. But, of course, there are a variety of data models which would work here.

I am on the fence about if it's better to use database-generated surrogate keys or UUIDs. For sure, if I'd used a UUID there (created in the Java program), this wouldn't have happened. But of course I was not aware of all this when I decided to go with a surrogate key in this case :)

> Think about the hour you wasted. This translates to several hours wasted on a regression

That's true enough about the change, a good point. And what makes this change even worse, is that at least with the TIMESTAMPTZ change my code no longer compiled, so I could clearly see there was work needing doing. This change would be much more subtle.

**Comment by lukaseder** [comment]
> An INSERT with everything default changes the data, but an UPDATE with nothing changed doesn't. So without triggers there is no need for a "no-change" UPDATE. But triggers are a thing...

Well, another side effect is the row lock, which can also be achieved with an equivalent `SELECT .. FOR UPDATE`, but I think that if the result set doesn't need to be consumed, an actual `UPDATE` would be more elegant.

> But in that case, the user table would just have one column (user_id) and the same problem would apply.

No, it would be worse. Your workarounds would not have worked, because there are no other dummy columns to set to `null` 😁

> Inevitably at one point or another this is going to leak through to the user due to some bug or oversight and they'll not understand why their username is some garbage string.

Or, they'll think their user name is `null`... Anyway, the actual use case is not the problem here. I think it's a good use-case to illustrate the general idea.

> > Think about the hour you wasted. This translates to several hours wasted on a regression
> 
> That's true enough about the change, a good point. And what makes this change even worse, is that at least with the TIMESTAMPTZ change my code no longer compiled, so I could clearly see there was work needing doing. This change would be much more subtle.

Well, to put things into perspective again, the change would not go unnoticed as easily if there's at least one `NOT NULL` column without any `DEFAULT` values, which is probably true for ~80% of all tables out there, so at least, there's an "obvious" regression in most systems.

The tricky case is when someone has a table like yours, with only PK and nullable columns.

**Comment by lukaseder** [comment]
The twitter poll was very conclusive:
https://twitter.com/lukaseder/status/1281600147575775232

Clearly, a very small minority favours the status quo with respect to `TableRecord.insert()`, the majority prefers for `insert()` to always run a query. With `update()`, it's not as clear. More people prefer the status quo where an `UPDATE` statement is skipped (as it is currently), but there are also a significant number of people who would like the update to be executed nonetheless.

I have created a new issue to introduce new `Settings` that govern the behaviour: https://github.com/jOOQ/jOOQ/issues/10386. Clearly, the default behaviour for `insert()` should be changed. The default behaviour for `update()` might remain the same as today - this requires a bit more thought.

This issue here will be implemented via #10386 and the new default for `Settings.insertUnchangedRecord`

**Comment by lukaseder** [comment]
This is now fixed via #10386, which changes the default for `TableRecord.insert()`. Remaining topics will be dealt with in #10386.
